File: negotiation-cordapp/workflows-java/src/test/java/negotiation/flows/AcceptanceFlowTests.java
package negotiation.flows;

import com.google.common.collect.ImmutableList;
import negotiation.states.ProposalState;
import negotiation.states.TradeState;
import net.corda.core.contracts.StateAndRef;
import net.corda.core.contracts.UniqueIdentifier;
import net.corda.core.identity.Party;
import org.junit.Assert;
import org.junit.Test;

import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

public class AcceptanceFlowTests extends FlowTestsBase{

    @Test
    public void acceptanceFlowConsumesTheProposalsInBothNodesVaultsAndReplacesWithEquivAccTradesWhenInitiatorIsBuyer() throws ExecutionException, InterruptedException {
        testAcceptance(true);

    }

    @Test
    public void acceptanceFlowConsumesTheProposalsInBothNodesVaultsAndReplacesWithEquivAccTradesWhenInitiatorIsSeller() throws ExecutionException, InterruptedException {
        testAcceptance(false);
    }

    @Test(expected = ExecutionException.class)
    public void acceptanceFlowThrowsAnErrorIfTheProposerTriesToAcceptTheProposal() throws ExecutionException, InterruptedException {
        int amount = 1;
        Party counterparty = b.getInfo().getLegalIdentitiesAndCerts().get(0).getParty();
        UniqueIdentifier proposalID = nodeACreatesProposal(true, amount, counterparty);

        AcceptanceFlow.Initiator flow = new AcceptanceFlow.Initiator(proposalID);
        Future future = a.startFlow(flow);
        network.runNetwork();
        future.get();
    }

    private void testAcceptance(Boolean isBuyer) throws ExecutionException, InterruptedException {
        int amount = 1;
        Party counterparty = b.getInfo().getLegalIdentitiesAndCerts().get(0).getParty();
        UniqueIdentifier proposalId = nodeACreatesProposal(isBuyer, amount, counterparty);
        nodeBAcceptsProposal(proposalId);
        ImmutableList.of(a,b).forEach(node -> {
            node.transaction(() -> {
                List<StateAndRef<ProposalState>> proposals = node.getServices().getVaultService().queryBy(ProposalState.class).getStates();
                Assert.assertEquals(0, proposals.size());

                List<StateAndRef<TradeState>> trades = node.getServices().getVaultService().queryBy(TradeState.class).getStates();
                Assert.assertEquals(1, trades.size());
                TradeState trade = trades.get(0).getState().getData();

                Assert.assertEquals(amount, trade.getAmount());
                Party buyer;
                Party seller;

                if(isBuyer){
                    buyer = a.getInfo().getLegalIdentitiesAndCerts().get(0).getParty();
                    seller = b.getInfo().getLegalIdentitiesAndCerts().get(0).getParty();
                }else{
                    seller = a.getInfo().getLegalIdentitiesAndCerts().get(0).getParty();
                    buyer = b.getInfo().getLegalIdentitiesAndCerts().get(0).getParty();
                }

                Assert.assertEquals(buyer, trade.getBuyer());
                Assert.assertEquals(seller, trade.getSeller());
                return null;
            });
            return;
        });
    }
}

File: negotiation-cordapp/workflows-kotlin/src/test/kotlin/negotiation/flows/flows/AcceptanceFlowTests.kt
package negotiation.flows.flows

import negotiation.flows.AcceptanceFlow
import negotiation.states.ProposalState
import negotiation.states.TradeState
import net.corda.core.flows.FlowException
import net.corda.core.node.services.queryBy
import net.corda.testing.internal.chooseIdentity
import org.junit.Test
import java.util.concurrent.ExecutionException
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class AcceptanceFlowTests: FlowTestsBase() {
    @Test
    fun `acceptance flow consumes the proposals in both nodes' vaults and replaces them with equivalent accepted trades when initiator is buyer`() {
        testAcceptance(false)
    }

    @Test
    fun `acceptance flow consumes the proposals in both nodes' vaults and replaces them with equivalent accepted trades when initiator is seller`() {
        testAcceptance(true)
    }

    @Test
    fun `acceptance flow throws an error is the proposer tries to accept the proposal`() {
        val amount = 1
        val counterparty = b.info.chooseIdentity()
        val proposalId = nodeACreatesProposal(true, amount, counterparty)

        val flow = AcceptanceFlow.Initiator(proposalId)
        val future = a.startFlow(flow)
        network.runNetwork()
        val exceptionFromFlow = assertFailsWith<ExecutionException> {
            future.get()
        }.cause!!
        assertEquals(FlowException::class, exceptionFromFlow::class)
        assertEquals("Only the proposee can accept a proposal.", exceptionFromFlow.message)
    }

    private fun testAcceptance(isBuyer: Boolean) {
        val amount = 1
        val counterparty = b.info.chooseIdentity()

        val proposalId = nodeACreatesProposal(isBuyer, amount, counterparty)
        nodeBAcceptsProposal(proposalId)

        for (node in listOf(a, b)) {
            node.transaction {
                val proposals = node.services.vaultService.queryBy<ProposalState>().states
                assertEquals(0, proposals.size)

                val trades = node.services.vaultService.queryBy<TradeState>().states
                assertEquals(1, trades.size)
                val trade = trades.single().state.data

                assertEquals(amount, trade.amount)
                val (buyer, seller) = when {
                    isBuyer -> listOf(a.info.chooseIdentity(), b.info.chooseIdentity())
                    else -> listOf(b.info.chooseIdentity(), a.info.chooseIdentity())
                }

                assertEquals(buyer, trade.buyer)
                assertEquals(seller, trade.seller)
            }
        }
    }
}